session/direct-ipc.c

```diff
@@ -0,0 +1,241 @@
+#define _POSIX_C_SOURCE 200809L
+#include <errno.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/sysmacros.h>
+#include <sys/wait.h>
+#include <xf86drm.h>
+#include <linux/major.h>
+#include <wlr/util/log.h>
+#include "session/direct-ipc.h"
+
+enum { DRM_MAJOR = 226 };
+
+#ifdef HAS_LIBCAP
+#include <sys/capability.h>
+
+static bool have_permissions(void) {
+	cap_t cap = cap_get_proc();
+	cap_flag_value_t val;
+
+	if (!cap || cap_get_flag(cap, CAP_SYS_ADMIN, CAP_PERMITTED, &val) || val != CAP_SET) {
```

Comment:
If you're root, you automatically get all of the caps anyway.
Checking for CAP_SYS_ADMIN explicitly means that while we still have some pretty serious permissions, we don't have free reign over the filesystem to be able to mess anything up.
Having just this cap and being part of the input group was enough for me to be able to run the examples.